File: frontend/lib/ambulances.ts
import { apiClient } from "@/lib/api-client";

export type Ambulance = {
  id: string;
  driver_name: string;
  vehicle_number: string;
  distance_km: number;
  eta_minutes: number;
  status: string;
  phone: string;
  ambulance_type: string;
};

export type AmbulanceBookingRequest = {
  patient_name: string;
  location: string;
  emergency_type: string;
  severity: string;
};

export type AmbulanceBookingResponse = {
  booking_id: string;
  assigned_ambulance: Ambulance;
  eta_minutes: number;
  status: "Dispatched";
  tracking_steps: string[];
};

export async function getAvailableAmbulances(): Promise<Ambulance[]> {
  return apiClient<Ambulance[]>("/api/ambulances/available", {
    errorMessage: "Unable to load available ambulances",
  });
}

export async function bookAmbulance(
  payload: AmbulanceBookingRequest,
): Promise<AmbulanceBookingResponse> {
  return apiClient<AmbulanceBookingResponse>("/api/ambulances/book", {
    errorMessage: "Unable to book ambulance",
    method: "POST",
    body: payload,
  });
}

File: backend/app/schemas/ambulances.py
from typing import Literal

from pydantic import BaseModel, Field


AmbulanceStatus = Literal["Available", "Busy", "Dispatched"]
BookingStatus = Literal["Dispatched"]


class Ambulance(BaseModel):
    id: str
    driver_name: str
    vehicle_number: str
    distance_km: float
    eta_minutes: int
    status: AmbulanceStatus
    phone: str
    ambulance_type: str


class AmbulanceBookingRequest(BaseModel):
    patient_name: str = Field(min_length=2, max_length=100)
    location: str = Field(min_length=3, max_length=200)
    emergency_type: str = Field(min_length=2, max_length=100)
    severity: str = Field(min_length=2, max_length=50)


class AmbulanceBookingResponse(BaseModel):
    booking_id: str
    assigned_ambulance: Ambulance
    eta_minutes: int
    status: BookingStatus
    tracking_steps: list[str]
